### PR TITLE
parser: implement Restore for PrivElem, GrantStmt and RevokeStmt

### PR DESCRIPTION
**Pull request by exialin**
### What problem does this PR solve? 
Implement `Restore` for `PrivElem`, `GrantStmt` and `RevokeStmt`.
Waiting for some helper `Restore`s implemented in #197 .

Issue: https://github.com/pingcap/tidb/issues/8532

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: ast/misc.go [hunk 1]
```diff
@@ -93,6 +93,18 @@ type AuthOption struct {
 	// TODO: support auth_plugin
 }
 
+// Restore implements Node interface.
+func (n *AuthOption) Restore(ctx *RestoreCtx) error {
```
Comment (tiancaiamao) [inline ast/misc.go]: Restore "IDENTIFIED WITH" will get "IDENTIFIED BY" ?

Comment (tiancaiamao) [reply ast/misc.go]: And there is no test to cover `AutoOption` ?

Comment (exialin) [reply ast/misc.go]: @tiancaiamao 
1. According to `parser.y`, it looks like syntax `IDENTIFIED WITH string` has no effect:
https://github.com/pingcap/parser/blob/63e2db7dbaddf946dc28c17077251b217fd5da26/parser.y#L7302-L7319

2. To test `Restore` of `AuthOption` (and `UserSpec`, `UserIdentity`), and to follow the testing routine of other types, it should be promoted as a `Node` and add `Accept` for it. Is that suitable?

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @leoppro

File: ast/misc.go [hunk 2]
```diff
@@ -975,7 +1111,40 @@ type GrantStmt struct {
...
+	}
+	ctx.WriteKeyWord(" ON ")
+	switch n.ObjectType {
+	case ObjectTypeNone:
+		// do nothing
+	case ObjectTypeTable:
+		ctx.WriteKeyWord("TABLE ")
+	default:
+		return errors.New("Unsupported object type of GrantStmt")
+	}
```
Comment (kennytm) [inline ast/misc.go]: Please implement `Restore` on `ObjectTypeType` to avoid duplicating this `switch` twice.

Comment (exialin) [reply ast/misc.go]: Modified. PTAL

File: ast/misc.go [hunk 3]
```diff
@@ -1250,6 +1308,19 @@ const (
 	ObjectTypeTable
 )
 
+// Restore implements Node interface.
+func (n ObjectTypeType) Restore(ctx *RestoreCtx) error {
+	switch n {
+	case ObjectTypeNone:
+		// do nothing
+	case ObjectTypeTable:
+		ctx.WriteKeyWord("TABLE ")
```
Comment (kennytm) [inline ast/misc.go]:
```suggestion
		ctx.WriteKeyWord("TABLE")
```

This also means line 1389 and 1447 would need to look like this 🤷‍♂️:

```go
if n.ObjectType != ObjectTypeNone {
	if err := n.ObjectType.Restore(ctx); err != nil {
		return errors.Annotate(err, "An error occurred while restore RevokeStmt.ObjectType")
	} 
	ctx.WritePlain(" ")
}
```

Comment (exialin) [reply ast/misc.go]: Yes, this is better. Should I create a new pull request to fix it? Sorry for the trouble.

Comment (kennytm) [reply ast/misc.go]: @exialin Yes please. Thanks.

**Pull request merged**